File: 프로그래머스/2/340212. ［PCCP 기출문제］ 2번 ／ 퍼즐 게임 챌린지/［PCCP 기출문제］ 2번 ／ 퍼즐 게임 챌린지.js
function solution(diffs, times, limit) { // 난이도, 소요시간, 전체 제한시간
    var answer = 0;
    
    let maxLev = 1;
    for (let i=0; i<diffs.length; i++){
        maxLev = Math.max(maxLev, diffs[i]);
    }
    
    const getPuzTime = (idx, level) => {
        if (diffs[idx]>level) return (diffs[idx]-level)*(times[idx]+(idx>0?times[idx-1]:0))+times[idx];
        else return times[idx];
    }
    
    const getTotTime = (level) => {
        let time = 0;
        for (let i=0; i<diffs.length; i++){
            time+=getPuzTime(i, level);
        }
        return time;
    }
    
    let start = 1;
    let end = maxLev;
    
    while (start<=end){
        let mid = Math.floor((start+end)/2);
        const time = getTotTime(mid);
        if (time<=limit){
            end = mid-1;
            answer = mid;
        } else {
            start = mid+1;
        }
    }
    
    return answer;
}

File: 프로그래머스/2/340212. ［PCCP 기출문제］ 2번 ／ 퍼즐 게임 챌린지/［PCCP 기출문제］ 2번 ／ 퍼즐 게임 챌린지.java
import java.util.*;

class Solution {
    public int solution(int[] diffs, int[] times, long limit) {
        int answer = 0;
        
        int maxLev = 1;
        for (int i=0; i<diffs.length; i++){
            maxLev = Math.max(maxLev, diffs[i]);
        }
        
        int start = 1; 
        int end = maxLev; 
        
        while (start <= end) {
            int mid = (start + end) / 2;
            long time = getTotTime(diffs, times, mid);
            if (time <= limit) {
                answer = mid; 
                end = mid - 1;  
            } else {
                start = mid + 1;
            }
        }
        
        return answer;
    }
    
    
    public int getPuzTime(int diff, int time_cur, int time_prev, int level) {
        if (diff>level) return (diff-level)*(time_cur+time_prev)+time_cur;
        else return time_cur;
    }
    
    public long getTotTime(int[] diffs, int[] times, int level){
        long time = 0;
        time+=getPuzTime(diffs[0], times[0], 0, level);
        for (int i=1; i<diffs.length; i++){
            time+=getPuzTime(diffs[i], times[i], times[i-1], level);
        }
        return time;
    }
}